### PR TITLE
[ONNX] Process const folding progressively when converts to ONNX

### PR DESCRIPTION
**Pull request by jiafatom**
This PR automatically solves onnx const attribute issue in PR https://github.com/pytorch/pytorch/pull/53784.

### CODE REVIEW

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 1]
```diff
@@ -650,7 +672,7 @@ def forward(self, x):
                 return x.new_zeros(x.data.size())
 
         x = torch.randn(3, 4)
-        self.run_test(Data(), x)
+        self.run_test(Data(), x, remained_onnx_input_idx=[])
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:675]: perhaps we should add `dynamic_axes` argument to these calls. That should preserve previous behavior

File: torch/onnx/utils.py [hunk 2]
```diff
@@ -1004,35 +1019,98 @@ def _run_symbolic_function(g, block, n, inputs, env, operator_export_type=Operat
...
+                    final_b_list = []
+                    for idx in range(len(if_output_list)):
+                        if current_b_list[idx] not in env:
+                            raise RuntimeError("The sub block ATen output " + current_b_list[idx] + " is not in env")
+                        if "onnx" in env[current_b_list[idx]].node().kind():
+                            onnx_b = env[current_b_list[idx]]
+                        else:
+                            onnx_b = g.op("Identity", env[current_b_list[idx]])
+                        final_b_list.append(onnx_b)
+                        torch._C._jit_pass_onnx_node_shape_type_inference(onnx_b.node(), _params_dict, opset_version,
```
Comment (BowenBao) [inline torch/onnx/utils.py:1078]: I see that you want to ensure shape inference was called, where output was not created by node inside this block. However, even if outside the block, the owner node should have shape inference called when being added to the graph right? The only exception might be graph input node (`prim::Param`), which we can probably do something special for it, like put its shape values into the map at initialization. Are there any other exceptions?

**Review by BowenBao — COMMENTED** [review]
Thanks for the huge PR, great to see this happening. I'm still going over the code, posting some current comments.

File: torch/onnx/symbolic_opset9.py [hunk 3]
```diff
@@ -1944,8 +1944,21 @@ def to(g, self, *args):
             # aten::to(Tensor, Device, bool, bool, memory_format)
             return self
         else:
-            dtype = sym_help._maybe_get_const(args[0], 'i')
-            if sym_help._is_value(dtype):
+            # TODO: The logic here is weird for test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime::test_ones_bool
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset9.py:1947]:
```
torch.to(other, non_blocking=False, copy=False) → Tensor

    Returns a Tensor with same torch.dtype and torch.device as the Tensor other.
    When non_blocking, tries to convert asynchronously with respect to the host if possible, 
    e.g., converting a CPU Tensor with pinned memory to a CUDA Tensor.
    When copy is set, a new Tensor is created even when the Tensor already matches the desired conversion.
```

This signature takes another tensor, and cast to the same dtype (and device).

File: torch/csrc/jit/passes/onnx/cast_all_constant_to_floating.cpp [hunk 4]
```diff
@@ -38,22 +40,23 @@ void CastAllConstantToFloating(Block* block) {
...
           default:
             throw std::runtime_error("Unsupported types: complex, string");
         }
         // create a cast node
         node->removeAttribute(attr::value);
-        node->t_(attr::value, val);
+        node->t_(attr::value, val_cast);
         Node* cast_node = graph->create(onnx::Cast, 1);
         cast_node->i_(attr::to, to_type);
+        cast_node->output()->setType(TensorType::create(val));
```
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/cast_all_constant_to_floating.cpp:59]: Is the creating of `val_cast` and `copy_` to avoid `TensorType::create(val)` being polluted? If so can we get the TensorType before cast and avoid the creating `val_cast` and `copy_`?

File: torch/csrc/jit/passes/onnx.cpp [hunk 5]
```diff
@@ -192,24 +204,34 @@ void BlockToONNX(
```
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx.cpp:204]: The comment is for the code above. when `is_sub_block==true`, shouldn't we avoid the above part adding input to `ctx.block`?

File: torch/csrc/jit/passes/onnx.cpp [hunk 6]
```diff
@@ -249,13 +271,33 @@ void NodeToONNX(
...
+          const_node->t_(attr::value, value);
+          const_node->output()->setType(TensorType::create(value));
+          // const_node->output()->setType(MergeInferredType(old->type(), outputs[i]->type()));
+
+          // Copy over source location and scope information to all nodes
+          // created by the symbolic
+          const_node->output()->node()->setSourceRange(node->sourceRange());
+          const_node->output()->node()->setScope(node->scope());
+          new_block->owningGraph()->appendNode(const_node);
+          ONNXShapeTypeInference(const_node, params_dict, opset_version, static_input_shape);
```
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx.cpp:290]: I see this is the source of introducing `params_dict` and `opset_version`. Can this be avoided by setting tensortype directly using api tensortype infer from tensor `value`?

File: torch/csrc/jit/passes/onnx.cpp [hunk 7]
```diff
@@ -249,13 +271,33 @@ void NodeToONNX(
         //
         // If onnx shape inference is turned on, the new outputs will have
         // types inferred, and they will be merged with the old types.
-        outputs[i]->setType(MergeInferredType(old->type(), outputs[i]->type()));
+        if (enable_const_folding &&
+            outputs[i]->node()->kind() != c10::onnx::Constant &&
+            ConstantValueMap::HasValue(outputs[i]->debugName())) {
+          // Perform const folding if the node output value is in ConstantValueMap.
```
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx.cpp:277]: nit: maybe rephrase this comment to be more elaborative. For once I thought it will perform constant folding like compute the actual value, but then realized the value was already computed, and are just fetched.

File: torch/csrc/jit/passes/onnx.cpp [hunk 8]
```diff
@@ -249,13 +271,33 @@ void NodeToONNX(
...
+          auto value =
+              ConstantValueMap::GetValue(outputs[i]->debugName()).value();
+          Node* const_node = new_block->owningGraph()->create(c10::onnx::Constant);
+          const_node->t_(attr::value, value);
+          const_node->output()->setType(TensorType::create(value));
+          // const_node->output()->setType(MergeInferredType(old->type(), outputs[i]->type()));
```
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx.cpp:283]: nit: remove comment

File: test/onnx/test_utility_funs.py [hunk 9]
```diff
@@ -35,6 +35,7 @@ def _model_to_graph(self, model, input,
                         training=TrainingMode.EVAL,
                         operator_export_type=OperatorExportTypes.ONNX):
 
+        _set_onnx_shape_inference(False)
```
[the remark this answers is not in the document]
Comment (jiafatom) [reply test/onnx/test_utility_funs.py:38]: sure, set back to True, and document the API as debug purpose.

This is the previous tested behavior. No need specify `dynamic_axes`. The reason is that our const folding technique rely on ConstantValueMap, which only gets effect when shape inference enabled. If shape inference is disabled, ConstantValueMap is never updated therefore no const folding happens.

File: test/onnx/test_onnx_opset.py [hunk 10]
```diff
@@ -186,13 +186,10 @@ class DynamicSliceModel(torch.jit.ScriptModule):
             def forward(self, x):
                 return x[1:x.size(0)]
```
[the remark this answers is not in the document]
Comment (jiafatom) [reply test/onnx/test_onnx_opset.py:189]: done.

File: test/onnx/test_onnx_opset.py [hunk 11]
```diff
@@ -282,36 +265,19 @@ def forward(self, x):
...
                   {"op_name" : "Concat"},
-                  {"op_name" : "Resize",
+                  {"op_name" : "Upsample",
                    "attributes" :
                    [{"name": "mode", "s": ("nearest").encode(), "type": 3}]}]
 
         ops = {9 : ops_9, 10 : ops_10}
         x = torch.randn(1, 2, 3, 4, requires_grad=True)
-        check_onnx_opsets_operator(MyModel(), x, ops, opset_versions=[9, 10])
+        check_onnx_opsets_operator(MyModel(), x, ops, opset_versions=[9])
```
[the remark this answers is not in the document]
Comment (jiafatom) [reply test/onnx/test_onnx_opset.py]: done.

File: test/onnx/test_pytorch_onnx_caffe2.py [hunk 12]
```diff
@@ -1177,7 +1181,7 @@ def forward(self, x, y):
 
         x = torch.randn(1, 2, requires_grad=True)
         y = torch.randn(2, 4, requires_grad=True)
-        self.run_model_test(MyModel(), train=False, input=(x, y), batch_size=BATCH_SIZE, use_gpu=False)
+        self.run_model_test(MyModel(), train=False, input=(x, y), batch_size=BATCH_SIZE, use_gpu=False, remained_onnx_input_idx=[0])
```
[the remark this answers is not in the document]
Comment (jiafatom) [reply test/onnx/test_pytorch_onnx_caffe2.py:1190]: done.

File: torch/csrc/jit/passes/onnx.cpp (hunk 7 above)
Comment (jiafatom) [reply torch/csrc/jit/passes/onnx.cpp]: Create a const node if the node output value is in ConstantValueMap.

File: torch/csrc/jit/passes/onnx.cpp (hunk 5 above)
Comment (jiafatom) [reply torch/csrc/jit/passes/onnx.cpp:215]: Added `!is_sub_block` to skip adding input to `ctx.block`

File: torch/csrc/jit/passes/onnx/cast_all_constant_to_floating.cpp (hunk 4 above)
Comment (jiafatom) [reply torch/csrc/jit/passes/onnx/cast_all_constant_to_floating.cpp]: done.

File: torch/onnx/symbolic_opset9.py (hunk 3 above)
Comment (jiafatom) [reply torch/onnx/symbolic_opset9.py]: The original code does not consider the case when `args[0]` is constant tensor (not int). This PR fixes it.

File: torch/onnx/utils.py [hunk 13]
```diff
@@ -1004,35 +1019,98 @@ def _run_symbolic_function(g, block, n, inputs, env, operator_export_type=Operat
...
-                            b_in.setType(inputs[i + 1].type())
-                    torch._C._jit_pass_onnx_block(b, new_block, operator_export_type, env)
-                new_op_outputs = torch._C._jit_pass_fixup_onnx_controlflow_node(new_node, opset_version)
-                # Process Loop and If after subblock is converted.
-                from torch.onnx.symbolic_helper import _onnx_shape_inference
-                if _onnx_shape_inference:
-                    torch._C._jit_pass_onnx_node_shape_type_inference(new_node, _params_dict, opset_version)
-                return new_op_outputs
+                    # env: 65:10, 21:14, 22:10, input.53:16, weight:17
+                    # embedding_matrix.1 and input.1 are prim::Param, so we need convert to ONNX Identity to do shape inference.
```
[the remark this answers is not in the document]
Comment (jiafatom) [reply torch/onnx/utils.py:1053]: This comments means that `prim::Param` (embedding_matrix.1 and input.1) is not in env, therefore we need convert to ONNX Identity to do shape inference. This also brings up a question when we handle `prim::Param`, shall we handle it in graph initialization? Will investigate.

File: torch/csrc/jit/passes/onnx.cpp [hunk 14]
```diff
@@ -163,20 +163,32 @@ void PreprocessCaffe2Ops(std::shared_ptr<Graph>& graph) {
...
   std::unordered_map<Value*, Value*> env;
-  BlockToONNX(graph->block(), new_graph->block(), operator_export_type, env);
+  auto enable_const_folding = true;
+  BlockToONNX(graph->block(), new_graph->block(), operator_export_type, env, static_input_shape, params_dict, opset_version, enable_const_folding);
   return new_graph;
 }
 
-void BlockToONNX(
+// BlockToONNX.
+// is_sub_block means the old_block (aten graph) is a sub block of the new block (onnx graph).
```
[the remark this answers is not in the document]
Comment (jiafatom) [reply torch/csrc/jit/passes/onnx.cpp]: Comments clarified:
```cpp
// is_sub_block = true means the old_block (aten graph) is in the sub block (e.g., if sub block),
// and we want to convert it into its parent block in onnx graph.
// In this case, we don't register the input/output or eliminate the dead code.
```

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 15]
```diff
@@ -1096,9 +1096,16 @@ void ProcessConstantValueMap(Node* n, int opset_version) {
...
         ConstantValueMap::SetRank(n->input(i)->debugName(), rank);
-        auto shape = input_type->symbolic_sizes();
-        if (!ConstantValueMap::HasShape(n->input(i)->debugName())) {
-          UpdateShape(n->input(i), shape);
+        // Only update shape if the input is onnx node.
+        // If it is aten operators, for example,
+        //   Float(20, 20, strides=[1, 0], requires_grad=0, device=cpu),
+        //     %399 : Float(20, 20, strides=[0, 1], requires_grad=0, device=cpu) = prim::ListUnpack(%397)
+        // The tracer shape may not be correct when dynamic_axes is enabled.
+        if (n->input(i)->node()->kind().is_onnx() || static_input_shape) {
```
[the remark this answers is not in the document]
Comment (jiafatom) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp:1116]: This is the only place. Now we extract this static info from graph by checking
```cpp
  for (auto n : g->inputs()) {
    if (!n->isCompleteTensor()) {
```
, rather than passing the argument all the way.

File: torch/_C/__init__.pyi.in [hunk 16]
```diff
@@ -308,14 +309,20 @@ def _jit_pass_onnx_eliminate_unused_items(graph: Graph, paramsDict: Dict[str, IV
...
     old_block: Block,
     new_block: Block,
     operator_export_type: _onnx.OperatorExportTypes,
-    env: Dict[Value, Value]
-) -> None: ...
+    env: Dict[Value, Value],
+    static_input_shape: _bool,
+    paramsDict: Dict[str, IValue],
+    opset_version: _int,
+    enable_const_folding: _bool,
```
[the remark this answers is not in the document]
Comment (jiafatom) [reply torch/_C/__init__.pyi.in:318]: This flag is actually in `_jit_pass_onnx_block` -- it is because this PR originally has a bug (will reply in your other comments).
Now it is not needed.

File: torch/csrc/jit/passes/onnx.cpp (hunk 6 above)
Comment (jiafatom) [reply torch/csrc/jit/passes/onnx.cpp]: We don't need pass `params_dict` and `opset_version`.
We create a const node here, in this case, `params_dict={}`. And we get `opset_version` through pybind.
The purpose of `ONNXShapeTypeInference` is used to update record in ConstantValueMap. For each generated ONNX operator (like const here), we need update ConstantValueMap.

File: torch/onnx/utils.py [hunk 17]
```diff
@@ -1004,35 +1019,98 @@ def _run_symbolic_function(g, block, n, inputs, env, operator_export_type=Operat
             elif op_name == "device" and n.output().type().kind() == "DeviceObjType":
                 return None
             elif op_name == 'Loop' or op_name == 'If':
-                new_op_outputs = g.op(op_name, *inputs, outputs=n.outputsSize())
-                new_node = new_op_outputs[0].node() if n.outputsSize() > 1 else new_op_outputs.node()
-                for b in n.blocks():
-                    new_block = new_node.addBlock()
-                    # Copy input metadata to subblock
+                static_if = False
+                enable_const_folding = False
```
[the remark this answers is not in the document]
Comment (jiafatom) [reply torch/onnx/utils.py:1007]: Now we don't need `enable_const_folding`.
Previously we need it - we cannot do const folding in if sub-block (nested if). - because this PR has a bug, that is:
When we create const node, we append it using `new_block->owningGraph()->appendNode(const_node);`
Then it will append the const node into the graph. This is not correct for sub-block case.
It triggers the test failure for `pytest test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime_opset11::test_floating_point_infer_dtype`
We should append it into the this node's corresponding block, e.g., `new_block->appendNode(const_node);`
After this fix, it is good. So we don't need  `enable_const_folding`.

File: torch/onnx/utils.py [hunk 18]
```diff
@@ -1004,35 +1019,98 @@ def _run_symbolic_function(g, block, n, inputs, env, operator_export_type=Operat
...
-                    torch._C._jit_pass_onnx_block(b, new_block, operator_export_type, env)
-                new_op_outputs = torch._C._jit_pass_fixup_onnx_controlflow_node(new_node, opset_version)
-                # Process Loop and If after subblock is converted.
-                from torch.onnx.symbolic_helper import _onnx_shape_inference
-                if _onnx_shape_inference:
-                    torch._C._jit_pass_onnx_node_shape_type_inference(new_node, _params_dict, opset_version)
-                return new_op_outputs
+                    # env: 65:10, 21:14, 22:10, input.53:16, weight:17
+                    # embedding_matrix.1 and input.1 are prim::Param, so we need convert to ONNX Identity to do shape inference.
+                    torch._C._jit_pass_onnx_node_shape_type_inference(inputs[0].node(), _params_dict, opset_version,
```
[the remark this answers is not in the document]
Comment (jiafatom) [reply torch/onnx/utils.py]: Right, no need shape inference here. This is const node and shape inference is done when const node is inserted. Remove shape inference here.

File: torch/onnx/utils.py [hunk 19]
```diff
@@ -1004,35 +1019,98 @@ def _run_symbolic_function(g, block, n, inputs, env, operator_export_type=Operat
...
+                                                                      get_static_input_shape())
+                    static_if = True
+                    input_flag = inputs[0].node()['value'].tolist()
```
[the remark this answers is not in the document]
Comment (jiafatom) [reply torch/onnx/utils.py:1038]: Right, as explained above, this shape inference is not needed.

File: torch/onnx/utils.py (hunk 2 above)
Comment (jiafatom) [reply torch/onnx/utils.py]: Right, this shape inference is not needed.  Shape inference is called when the node gets added to the graph, so it is already done. Remove this one.

File: torch/csrc/jit/passes/onnx/constant_fold.cpp [hunk 20]
```diff
@@ -299,7 +305,11 @@ c10::optional<at::Tensor> runTorchBackendForOnnx(
       std::vector<int64_t> axes;
       for (int64_t i = 0; i < inputTensorValues[1].sizes()[0]; ++i) {
         // ONNX unsqueeze accepts negative axes
-        axes_a[i] += axes_a[i] < 0 ? inputTensorValues[0].sizes().size() : 0;
+        // From https://pytorch.org/docs/stable/generated/torch.unsqueeze.html
+        // Negative dim will correspond to unsqueeze() applied at dim = dim +
+        // input.dim() + 1.
+        axes_a[i] +=
```
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/constant_fold.cpp:311]: wow, how come this was only discovered just now ....

File: torch/onnx/utils.py [hunk 21]
```diff
@@ -1004,35 +1002,88 @@ def _run_symbolic_function(g, block, n, inputs, env, operator_export_type=Operat
...
+                    current_b_list = list(current_b.outputs())
+
+                    final_b_list = []
+                    for idx in range(len(if_output_list)):
+                        if current_b_list[idx] not in env:
+                            raise RuntimeError("The sub block ATen output " + current_b_list[idx] + " is not in env")
+                        if "onnx" in env[current_b_list[idx]].node().kind():
+                            onnx_b = env[current_b_list[idx]]
+                        else:
+                            onnx_b = g.op("Identity", env[current_b_list[idx]])
```
Comment (BowenBao) [inline torch/onnx/utils.py:1053]: I think maybe you mentioned this before, could you remind me again why(for what case) this is needed?

Comment (jiafatom) [reply torch/onnx/utils.py]: Good point. This Identity operator is no longer needed, so I just remove this.
Previously it is because I need make Prim::Param node gets shape inference. In current logic, every onnx operator will trigger shape inference while inserting, so it won't need that.

**Review by BowenBao — APPROVED** [review]
LGTM, thanks!
